File: lab/backend/src/vm_lab/routers/displacement.py
"""POST /api/displacement — pairwise subpixel inter-frame shift over an ordered image
sequence (`vm.displacement`, `crates/vision-metrology/src/corr/displacement.rs`).

Each consecutive pair in `image_ids` is tracked independently (the same `window` rect,
taken from the *first* image of each pair — i.e. always in the previous frame's own
coordinates, not re-anchored to `image_ids[0]`); the response also reports the running
cumulative trajectory the Motion tab plots. A pair that comes back below `min_score` is
a 422, naming which pair failed, rather than silently dropping it from the trajectory.
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException

import vision_metrology as vm

from vm_lab.schemas import DisplacementPairOut, DisplacementRequest, DisplacementResponse
from vm_lab.store import store

router = APIRouter(prefix="/api/displacement", tags=["displacement"])


def _refine(req: DisplacementRequest) -> vm.Refine:
    if req.refine == "none":
        return vm.Refine.none()
    return vm.Refine.lucas_kanade(iters=req.lk_iters)


@router.post("", response_model=DisplacementResponse)
async def compute_displacement(req: DisplacementRequest) -> DisplacementResponse:
    for image_id in req.image_ids:
        if image_id not in store.images:
            raise HTTPException(404, f"no such image: {image_id}")

    cfg = vm.DisplacementConfig(
        window=req.window,
        search=(req.search_x, req.search_y),
        refine=_refine(req),
        min_score=req.min_score,
    )

    arrays = [store.load_array(image_id) for image_id in req.image_ids]

    pairs: list[DisplacementPairOut] = []
    cumulative_x = [0.0]
    cumulative_y = [0.0]
    for i in range(len(arrays) - 1):
        prev_id, curr_id = req.image_ids[i], req.image_ids[i + 1]
        try:
            d = vm.displacement(arrays[i], arrays[i + 1], cfg)
        except ValueError as exc:
            raise HTTPException(
                422, f"displacement failed between {prev_id} and {curr_id}: {exc}"
            ) from exc
        pairs.append(
            DisplacementPairOut(from_image_id=prev_id, to_image_id=curr_id, dx=d.dx, dy=d.dy, score=d.score)
        )
        cumulative_x.append(cumulative_x[-1] + d.dx)
        cumulative_y.append(cumulative_y[-1] + d.dy)

    return DisplacementResponse(pairs=pairs, cumulative_x=cumulative_x, cumulative_y=cumulative_y)

File: lab/frontend/src-tauri/src/commands/displacement.rs
//! `displacement` — mirrors `lab/backend/src/vm_lab/routers/displacement.py`.

use vision_metrology::corr::{DisplacementConfig, Refine, displacement as native_displacement};
use vm_primitives::Rect2f;

use crate::error::{AppResult, not_found};
use crate::state::AppState;
use crate::types::{DisplacementPairOut, DisplacementRequest, DisplacementResponse};

fn refine_from(req: &DisplacementRequest) -> Refine {
    if req.refine == "none" {
        Refine::None
    } else {
        Refine::LucasKanade {
            iters: req.lk_iters,
        }
    }
}

pub fn displacement(state: &AppState, req: DisplacementRequest) -> AppResult<DisplacementResponse> {
    let images = state.images.lock().expect("images mutex poisoned");
    for id in &req.image_ids {
        if !images.contains_key(id) {
            return Err(not_found("image", id));
        }
    }

    let cfg = DisplacementConfig {
        window: Rect2f {
            x: req.window[0],
            y: req.window[1],
            width: req.window[2],
            height: req.window[3],
        },
        search: (req.search_x, req.search_y),
        refine: refine_from(&req),
        min_score: req.min_score,
    };

    let mut pairs = Vec::with_capacity(req.image_ids.len().saturating_sub(1));
    let mut cumulative_x = vec![0.0f32];
    let mut cumulative_y = vec![0.0f32];
    for w in req.image_ids.windows(2) {
        let (prev_id, curr_id) = (&w[0], &w[1]);
        let prev = &images[prev_id].image;
        let curr = &images[curr_id].image;
        let d = native_displacement(&prev.as_view(), &curr.as_view(), &cfg).map_err(|e| {
            crate::error::AppError(format!(
                "displacement failed between {prev_id} and {curr_id}: {e}"
            ))
        })?;
        pairs.push(DisplacementPairOut {
            from_image_id: prev_id.clone(),
            to_image_id: curr_id.clone(),
            dx: d.shift.x,
            dy: d.shift.y,
            score: d.score,
        });
        cumulative_x.push(cumulative_x.last().unwrap() + d.shift.x);
        cumulative_y.push(cumulative_y.last().unwrap() + d.shift.y);
    }

    Ok(DisplacementResponse {
        pairs,
        cumulative_x,
        cumulative_y,
    })
}
